### PR TITLE
[Core] Missing await in Device.InvokeOnMainThreadAsync(Func<Task>)

### PR DESCRIPTION
**Pull request by dahlbyk**
### Description of Change ###

Stumbled on https://github.com/xamarin/Xamarin.Forms/pull/5028 as I was considering submitting an equivalent PR myself, and happened to notice a bug: `InvokeOnMainThreadAsync(Func<Task>)` doesn't `await`. I had originally just fixed the bug (note commit timestamps), but figured I might as well practice writing async tests.

**Update:** Apparently this was reported as #6705 and the fix in #6708 was merged an hour ago. Maybe the rest is still useful?

Also:
- Accepted VS suggestion to use local functions (microoptimization)
- Gave `dummyFunc` more useful names
- ~~Renamed `func`/`funcTask` to `function`, in alignment with `Task.Run(...)` overloads~~
  - Not sure if you consider renaming public arguments an API change? (It is.)

### API Changes ###

Changed:
 - ~~Names of some public arguments from `func`/`funcTask` to `function`~~

### Platforms Affected ### 

- Core/XAML (all platforms)

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

Not applicable

### Testing Procedure ###

Tried breaking the working code to hit each `Assert`.

### PR Checklist ###

- [x] Has automated tests
- [x] Rebased on top of the target branch at time of PR
- [ ] Changes adhere to coding standard

### CODE REVIEW

File: Xamarin.Forms.Core/Device.cs [hunk 1]
```diff
@@ -104,14 +104,14 @@ public static void BeginInvokeOnMainThread(Action action)
 			PlatformServices.BeginInvokeOnMainThread(action);
 		}
 
-		public static Task<T> InvokeOnMainThreadAsync<T>(Func<T> func)
+		public static Task<T> InvokeOnMainThreadAsync<T>(Func<T> function)
```
Comment (samhouts) [inline Xamarin.Forms.Core/Device.cs]: This rename is a breaking change

File: Xamarin.Forms.Core/Device.cs [hunk 2]
```diff
@@ -124,19 +124,19 @@ public static Task<T> InvokeOnMainThreadAsync<T>(Func<T> func)
...
 		public static Task InvokeOnMainThreadAsync(Action action)
 		{
-			Func<object> dummyFunc = () => { action(); return null; };
-			return InvokeOnMainThreadAsync(dummyFunc);
+			object wrapAction() { action(); return null; }
+			return InvokeOnMainThreadAsync(wrapAction);
 		}
 
-		public static Task<T> InvokeOnMainThreadAsync<T>(Func<Task<T>> funcTask)
+		public static Task<T> InvokeOnMainThreadAsync<T>(Func<Task<T>> function)
```
Comment (samhouts) [inline Xamarin.Forms.Core/Device.cs]: another break

File: Xamarin.Forms.Core/Device.cs [hunk 3]
```diff
@@ -149,10 +149,10 @@ public static Task<T> InvokeOnMainThreadAsync<T>(Func<Task<T>> funcTask)
 			return tcs.Task;
 		}
 
-		public static Task InvokeOnMainThreadAsync(Func<Task> funcTask)
+		public static Task InvokeOnMainThreadAsync(Func<Task> function)
```
Comment (samhouts) [inline Xamarin.Forms.Core/Device.cs]: break

File: Xamarin.Forms.Core/Device.cs (hunk 1 above)
Comment (dahlbyk) [reply Xamarin.Forms.Core/Device.cs]: Force pushed without these changes

**Comment by StephaneDelcroix** [comment]
> Update: Apparently this was reported as #6705 and the fix in #6708 was merged an hour ago. Maybe the rest is still useful?

the additional tests are useful

**Pull request merged**